compilers/concrete-compiler/compiler/lib/Dialect/TFHE/Transforms/TFHECircuitSolutionParametrization.cpp

```diff
@@ -203,15 +204,152 @@ class TFHEParametrizationTypeResolver : public TypeResolver {
       : solution(solution) {}
 
+  // Adds constraints that ensure that the argument types of `funcOp`
+  // and the operand types of `callOp` are equal and that the result
+  // types of all return operations of `funcOp` are equal to the
+  // result types of `callOp`
+  void addFuncAndCallConstraints(TypeConstraintSet<> &cs,
+                                 mlir::func::FuncOp funcOp,
+                                 mlir::func::CallOp callOp) {
+    // Ensure that the operands of the call op have the same
+    // type as the function arguments
+    for (auto [callOperand, funcArgument] :
+         llvm::zip_equal(callOp->getOperands(), funcOp.getArguments())) {
+      cs.addConstraint<DynamicSameTypeConstraint<DynamicFunctorYield>>(
+          [callOperand = callOperand]() { return callOperand; },
+          [funcArgument = funcArgument]() { return funcArgument; });
+    }
+
+    // Ensure that the return types of the function defined via
+    // its return ops are equal to the result types of the call
+    for (mlir::Block &block : funcOp.getRegion().getBlocks()) {
+      if (mlir::func::ReturnOp returnOp =
+              llvm::dyn_cast<mlir::func::ReturnOp>(block.getTerminator())) {
+        for (auto [returnOperand, callResult] :
+             llvm::zip_equal(returnOp->getOperands(), callOp->getResults())) {
+          cs.addConstraint<DynamicSameTypeConstraint<DynamicFunctorYield>>(
+              [returnOperand = returnOperand]() { return returnOperand; },
+              [callResult = callResult]() { return callResult; });
+        }
+      }
+    }
+  }
+
+  // Attempts to assemble the types for arguments and return values of
+  // the function with the name `funcName` from `state`. The (partial)
+  // result if passed in `retArgTypes` and `retResTypes`.
+  void
+  assembleInferredFunctionType(mlir::ModuleOp module, StringRef funcName,
+                               const LocalInferenceState &state,
+                               llvm::SmallVector<mlir::Type> &retArgTypes,
+                               llvm::SmallVector<mlir::Type> &retResTypes) {
+    mlir::func::FuncOp func = lookupFunction(module, funcName);
+    mlir::FunctionType funcType = func.getFunctionType();
+    llvm::SmallVector<mlir::Type> argTypes;
+    llvm::SmallVector<mlir::Type> resTypes(funcType.getNumResults());
+
+    // Assemble argument types
+    for (mlir::Value arg : func.getArguments()) {
+      argTypes.push_back(isUnresolvedType(arg.getType()) ? state.lookup(arg)
+                                                         : arg.getType());
+    }
+
+    // Assemble result types from types inferred for the operands
+    // of return ops
+    for (mlir::Block &block : func.getRegion().getBlocks()) {
+      if (mlir::func::ReturnOp returnOp =
+              llvm::dyn_cast<mlir::func::ReturnOp>(block.getTerminator()))
+        for (auto [i, operand] : llvm::enumerate(returnOp.getOperands())) {
+          if (!argTypes[i]) {
```

Comment:
The check and assignment should be on `resTypes` rather than `argTypes` here. This will be fixed in a fixup commit.

As far as the break on the first return op is concerned: the entire set of return types may be inferred from multiple return operations (e.g., the first one making it possible to infer type of the first result, the second one for the type of the second result), so breaking on the first one may leave the function type unresolved although its type could be inferred.

Regarding contradicting types: It is possible that type inference temporarily deduces two different types for the operand at the same position of two return ops. However, additional constraints verifying the equality of types ensure that the end result is coherent. Currently, this is only done indirectly via `func.call`, but should also be done with additional constraints directly in the rules for `func.func`. I will add these to a fixup commit.